src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,365 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.PatternMatching;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServices.Implementation.LanguageServiceBrokerShim;
+using Microsoft.VisualStudio.Text;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServices
+{
+    internal static class DocumentOutlineHelper
+    {
+        /// <summary>
+        /// The text snapshot from when the document symbol request was made.
+        /// </summary>
+        private static ITextSnapshot? LspSnapshot { get; set; }
+
+        /// <summary>
+        /// The text snapshot from the last active text view.
+        /// </summary>
+        private static ITextSnapshot? CurrentSnapshot { get; set; }
+
+        /// <summary>
+        /// Is true when LspSnapshot and CurrentSnapshot have been initialized.
+        /// </summary>
+        [MemberNotNullWhen(true, nameof(LspSnapshot), nameof(CurrentSnapshot))]
+        private static bool IsSnapshotInitialized { get; set; }
+
+        /// <summary>
+        /// The caret position of the last active text view.
+        /// </summary>
+        private static int CaretPosition { get; set; }
+
+        /// <summary>
+        /// Given an array of all Document Symbols in a document, returns an array containing the 
+        /// top-level Document Symbols and their nested children.
+        /// </summary>
+        public static DocumentSymbol[] GetNestedDocumentSymbols(DocumentSymbol[]? documentSymbols)
+        {
+            if (documentSymbols is null || documentSymbols.Length == 0)
+                return Array.Empty<DocumentSymbol>();
+
+            var allSymbols = documentSymbols
+                .SelectMany(x => x.Children)
```

Comment:
Can Children have further Children?

Reply:
No, that's why they need to be nested

Reply:
They CAN have further children but the LSP request unfortunately only returns 2 levels (a node and its children) and doesn't do any further nesting. That's why this code is necessary.